File: frontend/src/services/sanitize.ts
/**
 * Shared HTML-sanitization allowlist (single source of truth, frontend side).
 *
 * This MUST stay in sync with `backend/app/core/sanitize_policy.py`. The two
 * allowlists describe the same policy on each stack (nh3 on the backend,
 * DOMPurify here). Covers headings, lists, tables, code, images, and links.
 */
import DOMPurify from 'dompurify'

export const ALLOWED_TAGS: readonly string[] = [
  // headings
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  // text + structure
  'p',
  'br',
  'hr',
  'span',
  'div',
  'blockquote',
  'strong',
  'em',
  'b',
  'i',
  'u',
  's',
  'sub',
  'sup',
  // lists
  'ul',
  'ol',
  'li',
  // tables
  'table',
  'thead',
  'tbody',
  'tfoot',
  'tr',
  'th',
  'td',
  'caption',
  // code
  'code',
  'pre',
  // media + links
  'img',
  'a',
]

export const ALLOWED_ATTR: readonly string[] = [
  'class',
  'href',
  'title',
  'rel',
  'target',
  'src',
  'alt',
  'width',
  'height',
  'colspan',
  'rowspan',
  'scope',
]

/**
 * Inline images may only reference this app's own attachment-download endpoint.
 * MUST stay in sync with `_IMG_SRC` in `backend/app/core/sanitize.py`.
 */
export const IMG_SRC_PATTERN =
  /^\/api\/v1\/exercises\/[\w-]+\/reports\/[\w-]+\/attachments\/[\w-]+\/download$/

DOMPurify.addHook('afterSanitizeAttributes', (node) => {
  if (node.tagName === 'IMG' && !IMG_SRC_PATTERN.test(node.getAttribute('src') ?? '')) {
    node.removeAttribute('src')
  }
})

/** Sanitize untrusted HTML against the shared allowlist. */
export function sanitize(dirty: string): string {
  return DOMPurify.sanitize(dirty, {
    ALLOWED_TAGS: [...ALLOWED_TAGS],
    ALLOWED_ATTR: [...ALLOWED_ATTR],
  })
}

File: backend/app/core/sanitize.py
"""Server-authoritative HTML sanitization + plain-text extraction (nh3).

The writer editor (TipTap StarterKit) produces a constrained HTML subset.
``sanitize_html`` is the trust boundary for stored ``report_section.content``;
``html_to_plain`` derives ``content_plain`` / ``char_count`` for char-limit
enforcement and search.

The allowlist is NOT redefined here — it is sourced from
``app.core.sanitize_policy`` (the single source of truth, mirrored in
``frontend/src/services/sanitize.ts``). This module only wires nh3 to it.
"""

import re

import nh3

from app.core.sanitize_policy import ALLOWED_ATTRS, ALLOWED_TAGS

# nh3 wants plain ``set`` values and uses the "*" key for attributes allowed on
# every tag; the policy expresses that same global bucket with the "" key.
TIPTAP_ALLOWED: set[str] = set(ALLOWED_TAGS)
_NH3_ATTRS: dict[str, set[str]] = {("*" if tag == "" else tag): set(attrs) for tag, attrs in ALLOWED_ATTRS.items()}
_WS = re.compile(r"\s+")

# Inline images may only reference this app's own attachment-download endpoint
# (root-relative, fixed shape, dot-free segments — so no external/data:/traversal
# URLs survive sanitization). Mirrored in ``frontend/src/services/sanitize.ts``.
_IMG_SRC = re.compile(r"^/api/v1/exercises/[\w-]+/reports/[\w-]+/attachments/[\w-]+/download$")


def _attr_filter(tag: str, attr: str, value: str) -> str | None:
    if tag == "img" and attr == "src" and not _IMG_SRC.fullmatch(value):
        return None  # drop the attribute
    return value


def sanitize_html(html: str) -> str:
    """Return ``html`` reduced to the policy's allowed tag/attr set; scripts and JS URLs removed."""
    # link_rel=None: the policy owns the `rel` attribute (it is in the `a`
    # allowlist), so nh3 must not also guard/auto-inject it.
    return nh3.clean(html, tags=TIPTAP_ALLOWED, attributes=_NH3_ATTRS, link_rel=None, attribute_filter=_attr_filter)


def html_to_plain(html: str) -> str:
    """Strip every tag, unescape entities, collapse whitespace, trim."""
    stripped = nh3.clean(html, tags=set(), attributes={})
    return _WS.sub(" ", stripped).strip()
